x/transferinject/types/packet_metadata.go

```diff
@@ -0,0 +1,74 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	errortypes "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+// MemoData represents the structure of the memo with user and hub metadata
+type MemoData struct {
+	TransferInject *TransferInject `json:"transferinject"`
+}
+
+type TransferInject struct {
+	DenomMetadata *types.Metadata `json:"denom_metadata"`
+}
+
+func (p TransferInject) ValidateBasic() error {
+	return p.DenomMetadata.Validate()
+}
+
+const memoObjectKeyTransferInject = "transferinject"
+
+var (
+	ErrMemoTransferInjectEmpty         = fmt.Errorf("memo 'transferinject' is missing")
+	ErrMemoTransferInjectAlreadyExists = errorsmod.Wrapf(errortypes.ErrUnauthorized, "'transferinject' already exists in memo")
+)
+
+func ParsePacketMetadata(input string) (*TransferInject, error) {
+	bz := []byte(input)
+
+	var memo MemoData
```

Comment:
@zale144 why sometimes u unmarshal into `MemoData` and sometimes into `memoMap := make(map[string]any)`